### ISSUE TITLE
one-shot autocmd fails to close help window

### ISSUE THREAD

**Issue by lacygoill** [neovim/neovim#11308]
- `nvim --version`: NVIM v0.5.0-116-g4f4f5ef4f
 - `vim -u DEFAULTS` (version:  8.1 Included patches: 1-2231) behaves differently? Yes
 - Operating system/version: Ubuntu 16.04.6 LTS
 - Terminal name/version: XTerm(322)
 - `$TERM`: xterm-256color

### Steps to reproduce

Run this shell command:

    nvim -Nu NONE +'au QuickFixCmdPost * cw10|au CursorMoved * ++once helpc' +'helpg foobar'

### Actual behaviour

The help window is still opened.

### Expected behaviour

The help window is closed.

### Additional information

The same command works as expected in Vim.

---

If I move `:helpc` in a function and set a breakpoint into it, I can see that the function is called, and that `:helpc` is executed, but the help window is not closed (`winnr('$')` evaluates to `3` before and after `:helpc`):

    $ nvim -Nu <(cat <<'EOF'
    au QuickFixCmdPost * cw10|au CursorMoved * ++once call Func()
    breakadd func Func
    fu Func() abort
        helpc
    endfu
    EOF
    ) +'helpg foobar'

In Vim, `winnr('$')` evaluates to `3` before `:helpc`, but to `2` right after.

---

If I remove the `++once` argument, the help window is correctly closed, even without manually moving the cursor:

    nvim -Nu NONE +'au QuickFixCmdPost * cw10|au CursorMoved * helpc' +'helpg foobar'

---

If I replace `CursorMoved` with `BufWinEnter`, the issue is fixed:

    nvim -Nu NONE +'au QuickFixCmdPost * cw10|au BufWinEnter * helpc' +'helpg foobar'
                                                 ^^^^^^^^^^^

However, listening to `BufWinEnter` makes the cursor jump on the last line of the quickfix window.

Besides, listening to `BufWinEnter` may raise `E788` in Vim (in addition to `E480` which is expected):

    $ vim -Nu NONE +'au QuickFixCmdPost * cw10|au bufwinenter * ++once helpc' +'helpg foobar' +'helpg wont_find_this' +'helpg wont_find_this'

For some reason, the same command in Nvim does not raise `E788` (only `E480`):

    $ nvim -Nu NONE +'au QuickFixCmdPost * cw10|au bufwinenter * ++once helpc' +'helpg foobar' +'helpg wont_find_this' +'helpg wont_find_this'

---

Delaying `helpc` with a timer fixes the issue:

    nvim -Nu NONE +'au QuickFixCmdPost * cw10|au CursorMoved * call timer_start(0, { -> execute("helpc")})' +'helpg foobar'
                                                                    ^^^^^^^^^^^

---

I think `SafeState` would be the right event for a one-shot autocmd.
So for now, I'm going to use [this](https://github.com/lacygoill/vim-qf/blob/734f34995a63cfd279dfbcdb181a3a55e4fd1d23/autoload/qf.vim#L596-L601):

    if !has('nvim')
        au SafeState * ++once helpc
    else
        call timer_start(0, {-> execute('helpc')})
    endif

Hopefully, it will avoid all the pitfalls reported here.

**Comment by justinmk** [comment]
Maybe a missing vim patch ?

**Comment by lacygoill** [comment]
I guess it could be a missing Vim patch.
`++once` was added in Vim [8.1.1113](https://github.com/vim/vim/releases/tag/v8.1.1113).
I've just compiled this version, and the command already worked at the time.

But maybe the issue is not due to `++once`; or maybe it's affected by an older patch.
If it is, I'll try to find it later and report it here if I find anything.

**Comment by justinmk** [comment]
Thanks, I would look for help-related patches. Vim has tons of special-case code for the help window.

**Comment by lacygoill** [comment]
TLDR: The autocmd is probably an unreliable hack, so there is no guarantee that it works.
I think the most reliable event to delay `:helpclose` until the help window is opened after `:helpgrep` has been run, is `SafeState` (other events may lead to spurious errors); since it doesn't exist in Nvim, the best alternative is a timer with a 0ms waiting time.

---

I had a look at the Vim patches containing the keyword `help` in their commit message, back to when `:helpc` was introduced (in [`7.4.449`](https://github.com/vim/vim/releases/tag/v7.4.449)):

    8.1.0094  help text "usage:" is not capatalized
    8.1.0231  :help -? goes to help for -+
    8.1.0235  more help tags that jump to the wrong location
    8.1.0315  helpgrep with language doesn't work properly
    8.1.0505  filter command test may fail if helplang is not set
    8.1.0512  'helplang' default is inconsistent for C and C.UTF-8
    8.1.0541  help message in dosinst.c is outdated
    8.1.1226  {not in Vi} remarks get in the way of useful help text
    8.1.1280  remarks about functionality not in Vi clutters the help
    8.1.1629  terminal function help is in the wrong file
    8.1.1726  the eval.txt help file is too big
    8.1.1870  using :pedit from a help file sets help filetype
    8.1.2113  ":help expr-!~?" only works after searching

    8.0.0091  test_help_complete sometimes fails in MS-Windows console
    8.0.0116  using CTRl-] in English help language from 'helplang' is used
    8.0.0315  :help :[range] does not work
    8.0.0316  :help z? does not work
    8.0.0331  restoring help snapshot accesses freed memory
    8.0.0470  not enough testing for help commands
    8.0.0484  :lhelpgrep does not fail after a successful one
    8.0.0640  mismatch between help and actual message
    8.0.0649  when opening a help file the filetype is set several times
    8.0.0650  for extra help files the filetype is set more than once
    8.0.0704  problems with autocommands when opening help
    8.0.0729  the help for the terminal configure option is wrong
    8.0.1352  dead URLs in the help go unnoticed
    8.0.1383  local additions in help skips some files
    8.0.1384  not enough quickfix help; confusing winid
    8.0.1754  ex_helpgrep() is too long
    8.0.1792  MS-Windows users expect -? to work like --help

    7.4.017   ":help !!" does not find the "!!" tag in the help file
    7.4.104   ":help s/\_" reports an internal error
    7.4.170   some help tags don't work with ":help"
    7.4.449   can't easily close the help window

I compared this list to the one produced by `./scripts/vim-patch.sh -l`, which reduced the list to these patches:

    8.1.0315  helpgrep with language doesn't work properly
    8.1.1226  {not in Vi} remarks get in the way of useful help text
    8.1.1280  remarks about functionality not in Vi clutters the help
    8.1.1629  terminal function help is in the wrong file
    8.1.1726  the eval.txt help file is too big
    8.1.2113  ":help expr-!~?" only works after searching
    8.0.1754  ex_helpgrep() is too long

Out of those, only [`8.0.1754`](https://github.com/vim/vim/releases/tag/v8.0.1754) seem to be a possible candidate.

It's too old for `++once`, so I tested it with a regular autocmd in an augroup which clears itself:

    ./src/vim -Nu <(cat <<'EOF'
    set rtp=/usr/local/share/vim/vim81
    set hf=/usr/local/share/vim/vim81/doc/help.txt
    augroup my_helpc_later
        au!
        au QuickFixCmdPost * cw 10 | call My_helpc_later()
    augroup END
    fu My_helpc_later() abort
        augroup my_helpc_now
            au!
            au CursorMoved * helpc | au! my_helpc_now
        augroup END
    endfu
    helpg foobar
    EOF
    )

And it still worked as expected.
I then tried on `7.4.449`; same result, it already worked at the time.

**Comment by lacygoill** [comment]
> Just a guess, does reverting #9807 fix it? cc @mhinz who can explain it (PR and commits don't have an explanation :)

Indeed, you're right. I don't know how to revert a given commit, so I just compiled after checking out the commit https://github.com/neovim/neovim/commit/4c4a57015687df67c9bf48b721ab81e295f39118; the code fails; and on the parent commit (https://github.com/neovim/neovim/commit/11f03ee3f3da623440607220b8167a3b797d9daa), the code works.

So this commit explains the difference of behaviors; thank you for the explanation.
